[url_launcher] Migrate to the new embedding

mklim opened the pull request:
## Description

Migrate android intent to the new embedding.

- Refactors the existing plugin logic into two new classes.
- Adds an implementation of the plugin for the new embedding.
- Adds a simple unit test. It's not a very thorough one but does check that the new embedding's set up handlers work correctly.
- Adds a simple e2e test verifying that one of the getter APIs functions correctly.

## Related Issues

flutter/flutter#41849

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- x ] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

matthew-carroll commented on packages/url_launcher/url_launcher/android/src/main/java/dev/flutter/plugins/urllauncher/UrlLauncher.java:
```diff
@@ -0,0 +1,87 @@
+   *
+   * <p>It may be null initially, but should be set before calling {@link #launch}.
+   */
+  UrlLauncher(@Nullable Context activityContext) {
```
If you really need an `Activity` `Context` then I would recommend only accepting an `Activity`. It will enforce your requirements with the type system and avoid any future developer mistakes by overlooking the significance of `activityContext` vs `context`.

mklim replied:
Done. I took a second look at the original code, and I oversimplified. Half the time it was using an activity and half the time it was using application context. Split it out into separate objects and kept the `Activity` as that class following this suggestion.

matthew-carroll commented on packages/url_launcher/url_launcher/android/src/main/java/dev/flutter/plugins/urllauncher/UrlLauncherPlugin.java:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  @Override
+  public void onDetachedFromActivityForConfigChanges() {
```
Recommendation: consider declaring the methods in the order that they are likely to be called.

mklim replied:
Acknowledged. I prefer bottom line up front style (most "important" methods first), but it's a very subjective ranking.

matthew-carroll commented on packages/url_launcher/url_launcher/android/src/main/java/dev/flutter/plugins/urllauncher/UrlLauncherPlugin.java:
```diff
@@ -0,0 +1,58 @@
+ *
+ * <p>Instantiate this in an add to app scenario to gracefully handle activity and context changes.
+ */
+public final class UrlLauncherPlugin implements FlutterPlugin, ActivityAware {
```
Does this class need a static registerWith() method?

mklim replied:
I put it in a `UrlLauncherPluginRegistrar` class. I'm not 100% on that being the best place for it, but I wanted to move it as far away from here as possible to try and make it harder to accidentally initialize the "old" vs the "new" version of this.

mklim replied:
Never mind, our plugins infra expects Android and iOS to use the same name for the class that holds `registerWith`. I moved it back.

mklim commented:
This should now be ready to land. The latest commit updated the pubspec, added in the Gradle fix to keep this working on stable, and added a simple e2e test.

cyanglaz commented on packages/url_launcher/url_launcher/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncherPlugin.java:
```diff
@@ -1,110 +1,89 @@
+   *
+   * <p>See {@code dev.flutter.plugins.urllauncherexample.MainActivity} for an example.
+   */
+  public UrlLauncherPlugin() {}
```
Is this necessary?

mklim replied:
No, removed.

cyanglaz commented on packages/url_launcher/url_launcher/example/android/app/src/main/AndroidManifest.xml:
```diff
@@ -1,28 +1,41 @@
+    <activity
+      android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale|layoutDirection"
+      android:hardwareAccelerated="true"
+      android:launchMode="singleTop"
```
@matthew-carroll seems to suggest this to be moved.
https://github.com/flutter/plugins/pull/2164#discussion_r334189140

mklim replied:
Done.

cyanglaz commented:
also remember to upgrade the flutter version in example app.

mklim commented:
> also remember to upgrade the flutter version in example app.

Good catch, done.

amirh commented on packages/url_launcher/url_launcher/example/android/app/src/main/java/io/flutter/plugins/urllauncherexample/MainActivity.java:
```diff
@@ -1,18 +1,12 @@
-    super.onCreate(savedInstanceState);
-    GeneratedPluginRegistrant.registerWith(this);
+  public void configureFlutterEngine(FlutterEngine flutterEngine) {
+    flutterEngine.getPlugins().add(new UrlLauncherPlugin());
```
nit: we should have a TODO to remove this when the generated registrant for the v2 embedder is on stable
